### ISSUE TITLE
[Proposal] tRPC as communication between front and back

### ISSUE THREAD

**Issue by JoaquinTrinanes** [Vizzuality/front-end-scaffold#69]
_(I'm opening this as an issue because we don't have discussions enabled)_

# The rant

There are a LOT of things that can go wrong when communicating between front and back through an API. One of the most commons issue I've encountered was the typing system and schema validation. We can send whatever we want to whatever endpoint, and it's a responsibility of both the front and back team to coordinate and determine this.
There are many popular solutions for this:
- **OpenAPI specification.** That is, a file that lives outside of the code and sometimes the project. It instructs the available endpoints and how to communicate with them. Sometimes it's generated automatically using tools that analyse the existing APIs, but if the feature to develop is new it would block the front team until it's completed, so it should be written beforehand. This method is beloved by JS and Python developers because it allows the developer to break the contract without consequences (during linting or build time) (we use this! 🎉 ) 
- **GraphQL.** What if we host the specification on a server, version it and create a typed wrapper around it? GraphQL is the standard de-facto for type-safe API consumption, but it implies a huge developer overhead (also the schema has to be synched every time you change it)

During the past weeks, I had to do a HUGE change in one of my Next.js personal projects. What was originally a static site that read data from a CMS turned into an application that also provided CRUD access to a database, authorization, magic link sign-in, user roles...

I felt a little hopeless at first, as I don't really enjoy creating a file per Next endpoint, and then, per each file, add the following:
1. Guards to avoid calling a not allowed method
  ```ts
  switch (req.method) {
    case 'GET':
        ...
    default:
        return res.status(405);
  ```
2. Type validation, to ensure the body I received (of type `unkown`) was correct. That validation can be different depending on the HTTP method.
3. When consuming the API, cast the response type to what I expected the endpoint to return.
  ```ts
      return response.data as ResponseInterventionsData; // this won't fail if the response type is different
  ```

# Current situation
We don't use Next.js endpoint's at Vizz that I know of (except for specific cases like authentication), but the problems described above exist consuming an external API too. The backend team has to create a DTO to validate the data, the frontend team has to cast whatever arrives to something usable. And there's a Swagger somewhere (that being said, in comparison to my other jobs, you do a great work being faithful to the specification, even though I found some documentation without response type in the wild that I had to check by hand).

When the specification changes, the backend has to change both the spec and the DTOs. The front has to modify the interfaces they created to do their castings (usually from the spec).

There's got to be a better way, right? RIGHT?

# Enter [tRPC](https://trpc.io/)

The thing I like about tRPC is the first thing that appears when you open the page
![image](https://storage.googleapis.com/trpc/trpcgif.gif)

In a nutshell, tRPC provides **inferred, type safe Remote Procedure Calls**.

This is the workflow when creating a tRPC method (I'll use v9 syntax, v10 is on the way but still in beta):
1. Add the call to the router:
  ```ts
  ```
